util/global_conn_id.go

```diff
@@ -0,0 +1,453 @@
+	}
+	return GlobalConnID{
+		Is64bits:    false,
+		LocalConnID: (id >> 1) & MaxLocalConnID32,
+		ServerID:    (id >> 21) & MaxServerID32,
+	}, false, nil
+}
+
+// GlobalConnIDAllocator is global connection ID allocator.
+type GlobalConnIDAllocator struct {
+	is64bits       sync2.AtomicInt32 // !0: true, 0: false
+	serverIDGetter func() uint64
+
+	local32 LocalConnIDAllocator32
+	local64 LocalConnIDAllocator64
+}
+
+// Is64 indicates allocate 64bits global connection ID or not.
+func (g *GlobalConnIDAllocator) Is64() bool {
+	return g.is64bits.Get() != 0
+}
+
+// UpgradeTo64 upgrade allocator to 64bits.
+func (g *GlobalConnIDAllocator) UpgradeTo64() {
+	g.is64bits.Set(1)
+}
+
+// Init initiate members.
+func (g *GlobalConnIDAllocator) Init(serverIDGetter serverIDGetterFn, existedChecker connectionIDExistCheckerFn) {
+	g.serverIDGetter = serverIDGetter
+
+	g.local32.Init(nil)
+	g.local64.Init(existedChecker)
+
+	g.is64bits.Set(1) // TODO: set 32bits as default, after 32bits logics is fully implemented and tested.
+}
+
+// SetServerIDGetter set serverIDGetter member.
+func (g *GlobalConnIDAllocator) SetServerIDGetter(serverIDGetter serverIDGetterFn) {
+	g.serverIDGetter = serverIDGetter
+}
+
+// NextID returns next connection ID.
+func (g *GlobalConnIDAllocator) NextID() uint64 {
+	globalConnID := g.Allocate()
+	return globalConnID.ID()
+}
+
+// Allocate allocates a new global connID.
+func (g *GlobalConnIDAllocator) Allocate() GlobalConnID {
+	serverID := g.serverIDGetter()
+
+	// 32bits.
+	if !g.Is64() {
+		localConnID, ok := g.local32.Allocate()
+		if ok {
+			return GlobalConnID{
+				ServerID:    serverID,
+				LocalConnID: localConnID,
+				Is64bits:    false,
+			}
+		}
+		g.UpgradeTo64() // go on to 64bits.
+	}
+
+	// 64bits.
+	return GlobalConnID{
+		ServerID:    serverID,
+		LocalConnID: g.local64.Allocate(serverID),
+		Is64bits:    true,
+	}
+}
+
+// Release releases connectionID to pool.
+func (g *GlobalConnIDAllocator) Release(connectionID uint64) {
+	globalConnID, isTruncated, err := ParseGlobalConnID(connectionID)
+	if err != nil || isTruncated {
+		logutil.BgLogger().Error("failed to ParseGlobalConnID", zap.Error(err), zap.Uint64("connectionID", connectionID), zap.Bool("isTruncated", isTruncated))
+		return
+	}
+
+	if globalConnID.Is64bits {
+		g.local64.Deallocate(globalConnID.LocalConnID)
+	} else {
+		if err = g.local32.Deallocate(globalConnID.LocalConnID); err != nil {
+			logutil.BgLogger().Error("failed to release 32bits connection ID", zap.Error(err), zap.Uint64("connectionID", connectionID), zap.Uint64("localConnID", globalConnID.LocalConnID))
+		}
+	}
+}
+
+// LocalConnIDAllocator64 is local connID allocator for 64bits global connection ID.
+type LocalConnIDAllocator64 struct {
+	existedChecker connectionIDExistCheckerFn
+	lastID         uint64
+}
+
+// Init initiates LocalConnIDAllocator64
+func (a *LocalConnIDAllocator64) Init(existedChecker connectionIDExistCheckerFn) {
+	a.existedChecker = existedChecker
```

Comment:
How about maintaining the exist checker inside the allocator, considering that it provides alloc and dealloc so that it can fully know whether ID exists?

Reply:
Updated. PTAL~